### PR TITLE
♻️ Leverage 'Point/Displacement' in more places

### PR DESCRIPTION
**Pull request by julealgon**

### CODE REVIEW

**Comment by julealgon** [comment]
Depends on https://github.com/diasurgical/devilutionX/pull/2788

**Review by StephenCWills — DISMISSED** [review]

File: Source/loadsave.cpp [hunk 1]
```diff
@@ -757,8 +757,7 @@ void LoadQuest(LoadHelper *file, int i)
 		file->Skip(1); // Alignment
 	quest._qlog = file->NextBool32();
 
-	ReturnLvlX = file->NextBE<int32_t>();
-	ReturnLvlY = file->NextBE<int32_t>();
+	ReturnLvlPosition = { file->NextBE<int32_t>(), file->NextBE<int32_t>() };
```
[the remark this answers is not in the document]
Comment (AJenbo) [reply Source/loadsave.cpp:760]: I think recent versions of VS had a similar issue.

[the remark this answers is not in the document]
Comment (StephenCWills) [reply Source/loadsave.cpp:760]: Unfortunately, I consider that to be fairly believable. I'd certainly feel more comfortable merging this myself if the order was made explicit like the `AnimInfo` example.

File: Source/scrollrt.cpp [hunk 2]
```diff
@@ -557,47 +549,43 @@ void DrawPlayer(const Surface &out, int pnum, int x, int y, int px, int py)
...
 		auto &player = Players[i];
-		if (player.plractive && player._pHitPoints == 0 && player.plrlevel == (BYTE)currlevel && player.position.tile.x == x && player.position.tile.y == y) {
-			dFlags[x][y] |= BFLAG_DEAD_PLAYER;
-			int px = sx + player.position.offset.deltaX - CalculateWidth2(player.AnimInfo.pCelSprite == nullptr ? 96 : player.AnimInfo.pCelSprite->Width());
-			int py = sy + player.position.offset.deltaY;
-			DrawPlayer(out, i, x, y, px, py);
+		if (player.plractive && player._pHitPoints == 0 && player.plrlevel == (BYTE)currlevel && player.position.tile == tilePosition) {
+			dFlags[tilePosition.x][tilePosition.y] |= BFLAG_DEAD_PLAYER;
+			const int width { CalculateWidth2(player.AnimInfo.pCelSprite == nullptr ? 96 : player.AnimInfo.pCelSprite->Width()) };
+			const Point playerRenderPosition { targetBufferPosition + player.position.offset - Displacement { width, 0 } };
```
Comment (AJenbo) [inline Source/scrollrt.cpp:573]:
```suggestion
			const Displacement center { CalculateWidth2(player.AnimInfo.pCelSprite == nullptr ? 96 : player.AnimInfo.pCelSprite->Width()), 0 };
			const Point playerRenderPosition { targetBufferPosition + player.position.offset - center };
```

File: Source/scrollrt.cpp [hunk 3]
```diff
@@ -826,57 +802,56 @@ void DrawPlayerHelper(const Surface &out, int x, int y, int sx, int sy)
 	if (player.IsWalking()) {
 		offset = GetOffsetForWalking(player.AnimInfo, player._pdir);
 	}
-	int px = sx + offset.deltaX - CalculateWidth2(player.AnimInfo.pCelSprite == nullptr ? 96 : player.AnimInfo.pCelSprite->Width());
-	int py = sy + offset.deltaY;
 
-	DrawPlayer(out, p, x, y, px, py);
+	const int width { CalculateWidth2(player.AnimInfo.pCelSprite == nullptr ? 96 : player.AnimInfo.pCelSprite->Width()) };
+	const Point playerRenderPosition { targetBufferPosition + offset - Displacement { width, 0 } };
```
Comment (AJenbo) [inline Source/scrollrt.cpp:807]:
```suggestion
	const Displacement center { CalculateWidth2(player.AnimInfo.pCelSprite == nullptr ? 96 : player.AnimInfo.pCelSprite->Width()), 0 };
	const Point playerRenderPosition { targetBufferPosition + offset - center };

```

File: Source/scrollrt.cpp [hunk 4]
```diff
@@ -948,45 +923,43 @@ void DrawDungeon(const Surface &out, int sx, int sy, int dx, int dy)
...
-					world_draw_black_tile(out, sx, sy);
+					world_draw_black_tile(out, targetBufferPosition.x, targetBufferPosition.y);
 				}
 			} else {
-				world_draw_black_tile(out, sx, sy);
+				world_draw_black_tile(out, targetBufferPosition.x, targetBufferPosition.y);
 			}
-			ShiftGrid(&x, &y, 1, 0);
-			sx += TILE_WIDTH;
+			ShiftGrid(&tilePosition.x, &tilePosition.y, 1, 0);
```
Comment (AJenbo) [inline Source/scrollrt.cpp:946]:
```suggestion
			tilePosition += DIR_E;
```

File: Source/scrollrt.cpp [hunk 5]
```diff
@@ -997,54 +970,52 @@ void DrawFloor(const Surface &out, int x, int y, int sx, int sy, int rows, int c
...
 				}
-				if (dPiece[x][y] != 0) {
-					DrawDungeon(out, x, y, sx, sy);
+				if (dPiece[tilePosition.x][tilePosition.y] != 0) {
+					DrawDungeon(out, tilePosition, targetBufferPosition);
 				}
 			}
-			ShiftGrid(&x, &y, 1, 0);
-			sx += TILE_WIDTH;
+			ShiftGrid(&tilePosition.x, &tilePosition.y, 1, 0);
```
Comment (AJenbo) [inline Source/scrollrt.cpp:1002]:
```suggestion
			tilePosition += DIR_E;
```

File: Source/scrollrt.cpp [hunk 6]
```diff
@@ -1129,36 +1097,35 @@ void DrawGame(const Surface &fullOut, int x, int y)
...
 			if (invflag || sbookflag) {
-				ShiftGrid(&x, &y, 2, 0);
+				ShiftGrid(&position.x, &position.y, 2, 0);
 				columns -= 4;
 				sx += -TILE_WIDTH / 2;
 			}
 		} else {
 			if (chrflag || QuestLogIsOpen) {
-				ShiftGrid(&x, &y, 1, 0);
+				ShiftGrid(&position.x, &position.y, 1, 0);
```
Comment (AJenbo) [inline Source/scrollrt.cpp:1123]:
```suggestion
				position += DIR_E;
```

File: Source/scrollrt.cpp [hunk 7]
```diff
@@ -1129,36 +1097,35 @@ void DrawGame(const Surface &fullOut, int x, int y)
...
 				columns -= 2;
 				sx += -TILE_WIDTH / 2 / 2; // SPANEL_WIDTH accounted for in Zoom()
 			}
 			if (invflag || sbookflag) {
-				ShiftGrid(&x, &y, 1, 0);
+				ShiftGrid(&position.x, &position.y, 1, 0);
```
Comment (AJenbo) [inline Source/scrollrt.cpp:1128]:
```suggestion
				position += DIR_E;
```

File: Source/scrollrt.cpp [hunk 8]
```diff
@@ -1192,28 +1159,28 @@ void DrawGame(const Surface &fullOut, int x, int y)
...
 		sx -= TILE_WIDTH;
-		ShiftGrid(&x, &y, -1, 0);
+		ShiftGrid(&position.x, &position.y, -1, 0);
 		columns++;
 		break;
 	case SDIR_NW:
 		sx -= TILE_WIDTH / 2;
 		sy -= TILE_HEIGHT / 2;
-		x--;
+		position.x--;
```
Comment (AJenbo) [inline Source/scrollrt.cpp:1174]:
```suggestion
		position += DIR_NW;
```

File: Source/scrollrt.cpp [hunk 9]
```diff
@@ -1171,12 +1138,12 @@ void DrawGame(const Surface &fullOut, int x, int y)
 	switch (ScrollInfo._sdir) {
 	case SDIR_N:
 		sy -= TILE_HEIGHT;
-		ShiftGrid(&x, &y, 0, -1);
+		ShiftGrid(&position.x, &position.y, 0, -1);
```
Comment (AJenbo) [inline Source/scrollrt.cpp:1141]:
```suggestion
		position += DIR_N;
```

File: Source/scrollrt.cpp [hunk 10]
```diff
@@ -1171,12 +1138,12 @@ void DrawGame(const Surface &fullOut, int x, int y)
...
 	case SDIR_N:
 		sy -= TILE_HEIGHT;
-		ShiftGrid(&x, &y, 0, -1);
+		ShiftGrid(&position.x, &position.y, 0, -1);
 		rows += 2;
 		break;
 	case SDIR_NE:
 		sy -= TILE_HEIGHT;
-		ShiftGrid(&x, &y, 0, -1);
+		ShiftGrid(&position.x, &position.y, 0, -1);
```
Comment (AJenbo) [inline Source/scrollrt.cpp:1146]:
```suggestion
		position += DIR_N;
```

File: Source/scrollrt.cpp [hunk 11]
```diff
@@ -1192,28 +1159,28 @@ void DrawGame(const Surface &fullOut, int x, int y)
 		break;
 	case SDIR_SW:
 		sx -= TILE_WIDTH;
-		ShiftGrid(&x, &y, -1, 0);
+		ShiftGrid(&position.x, &position.y, -1, 0);
```
Comment (AJenbo) [inline Source/scrollrt.cpp:1162]:
```suggestion
		position += DIR_W;
```

File: Source/scrollrt.cpp [hunk 12]
```diff
@@ -1192,28 +1159,28 @@ void DrawGame(const Surface &fullOut, int x, int y)
...
 		sx -= TILE_WIDTH;
-		ShiftGrid(&x, &y, -1, 0);
+		ShiftGrid(&position.x, &position.y, -1, 0);
 		columns++;
 		rows++;
 		break;
 	case SDIR_W:
 		sx -= TILE_WIDTH;
-		ShiftGrid(&x, &y, -1, 0);
+		ShiftGrid(&position.x, &position.y, -1, 0);
```
Comment (AJenbo) [inline Source/scrollrt.cpp:1168]:
```suggestion
		position += DIR_W;
```

File: Source/scrollrt.cpp [hunk 13]
```diff
@@ -1129,36 +1097,35 @@ void DrawGame(const Surface &fullOut, int x, int y)
...
-	x += tileShiftX;
-	y += tileShiftY;
+	position += tileShift;
 
 	// Skip rendering parts covered by the panels
 	if (CanPanelsCoverView()) {
 		if (zoomflag) {
 			if (chrflag || QuestLogIsOpen) {
-				ShiftGrid(&x, &y, 2, 0);
+				ShiftGrid(&position.x, &position.y, 2, 0);
```
Comment (AJenbo) [inline Source/scrollrt.cpp:1112]:
```suggestion
				position += DIR_E;
				position += DIR_E;
```

File: Source/scrollrt.cpp [hunk 14]
```diff
@@ -1129,36 +1097,35 @@ void DrawGame(const Surface &fullOut, int x, int y)
...
 				columns -= 4;
 				sx += SPANEL_WIDTH - TILE_WIDTH / 2;
 			}
 			if (invflag || sbookflag) {
-				ShiftGrid(&x, &y, 2, 0);
+				ShiftGrid(&position.x, &position.y, 2, 0);
```
Comment (AJenbo) [inline Source/scrollrt.cpp:1117]:
```suggestion
				position += DIR_E;
				position += DIR_E;
```

File: Source/scrollrt.cpp [hunk 15]
```diff
@@ -997,54 +970,52 @@ void DrawFloor(const Surface &out, int x, int y, int sx, int sy, int rows, int c
...
 			}
-			ShiftGrid(&x, &y, 1, 0);
-			sx += TILE_WIDTH;
+			ShiftGrid(&tilePosition.x, &tilePosition.y, 1, 0);
+			targetBufferPosition.x += TILE_WIDTH;
 		}
 		// Return to start of row
-		ShiftGrid(&x, &y, -columns, 0);
-		sx -= columns * TILE_WIDTH;
+		ShiftGrid(&tilePosition.x, &tilePosition.y, -columns, 0);
```
Comment (AJenbo) [inline Source/scrollrt.cpp:1006]:
```suggestion
		for (int i = 0; i < columns; 1++)
			tilePosition += DIR_W;
```

File: Source/scrollrt.cpp [hunk 16]
```diff
@@ -1552,42 +1518,40 @@ void TilesInView(int *rcolumns, int *rrows)
...
 		if ((lrow & 1) == 0) {
 			// Offset tile to vertically align the player when both rows and colums are even
 			tileRows++;
-			tileOffsetY -= TILE_HEIGHT / 2;
+			tileOffset.deltaY -= TILE_HEIGHT / 2;
 		}
 	} else if ((tileColums & 1) != 0 && (lrow & 1) != 0) {
 		// Offset tile to vertically align the player when both rows and colums are odd
-		ShiftGrid(&tileShiftX, &tileShiftY, 0, -1);
+		ShiftGrid(&tileShift.deltaX, &tileShift.deltaY, 0, -1);
```
Comment (AJenbo) [inline Source/scrollrt.cpp:1547]:
```suggestion
		tileShift += DIR_W;
```

File: Source/scrollrt.cpp [hunk 17]
```diff
@@ -1552,42 +1518,40 @@ void TilesInView(int *rcolumns, int *rrows)
...
-	tileOffsetX = 0 - xo;
-	tileOffsetY = 0 - yo - 1 + TILE_HEIGHT / 2;
+	tileOffset = { -xo, -yo - 1 + TILE_HEIGHT / 2 };
 
 	TilesInView(&tileColums, &tileRows);
 	int lrow = tileRows - RowsCoveredByPanel();
 
 	// Center player tile on screen
-	ShiftGrid(&tileShiftX, &tileShiftY, -tileColums / 2, -lrow / 2);
+	ShiftGrid(&tileShift.deltaX, &tileShift.deltaY, -tileColums / 2, -lrow / 2);
```
Comment (AJenbo) [inline Source/scrollrt.cpp:1533]:
```suggestion
	tileShift += Displacement { DIR_W } * (tileColums / 2);
	tileShift += Displacement { DIR_N } * (lrow / 2);
```

File: Source/scrollrt.cpp [hunk 18]
```diff
@@ -948,45 +923,43 @@ void DrawDungeon(const Surface &out, int sx, int sy, int dx, int dy)
...
 			}
-			ShiftGrid(&x, &y, 1, 0);
-			sx += TILE_WIDTH;
+			ShiftGrid(&tilePosition.x, &tilePosition.y, 1, 0);
+			targetBufferPosition.x += TILE_WIDTH;
 		}
 		// Return to start of row
-		ShiftGrid(&x, &y, -columns, 0);
-		sx -= columns * TILE_WIDTH;
+		ShiftGrid(&tilePosition.x, &tilePosition.y, -columns, 0);
```
Comment (AJenbo) [inline Source/scrollrt.cpp:950]:
```suggestion
		tilePosition += Displacement { DIR_W } * columns;
```

File: Source/scrollrt.cpp [hunk 19]
```diff
@@ -997,54 +970,52 @@ void DrawFloor(const Surface &out, int x, int y, int sx, int sy, int rows, int c
...
 					// Render objects behind walls first to prevent sprites, that are moving
 					// between tiles, from poking through the walls as they exceed the tile bounds.
 					// A proper fix for this would probably be to layout the sceen and render by
 					// sprite screen position rather than tile position.
-					if (IsWall(x, y) && (IsWall(x + 1, y) || (x > 0 && IsWall(x - 1, y)))) { // Part of a wall aligned on the x-axis
-						if (IsWalkable(x + 1, y - 1) && IsWalkable(x, y - 1)) {              // Has walkable area behind it
-							DrawDungeon(out, x + 1, y - 1, sx + TILE_WIDTH, sy);
+					if (IsWall(tilePosition.x, tilePosition.y) && (IsWall(tilePosition.x + 1, tilePosition.y) || (tilePosition.x > 0 && IsWall(tilePosition.x - 1, tilePosition.y)))) { // Part of a wall aligned on the x-axis
+						if (IsWalkable(tilePosition.x + 1, tilePosition.y - 1) && IsWalkable(tilePosition.x, tilePosition.y - 1)) {                                                     // Has walkable area behind it
+							DrawDungeon(out, tilePosition + Displacement { 1, -1 }, { targetBufferPosition.x + TILE_WIDTH, targetBufferPosition.y });
```
Comment (AJenbo) [inline Source/scrollrt.cpp:994]:
```suggestion
							DrawDungeon(out, tilePosition + DIR_E, { targetBufferPosition.x + TILE_WIDTH, targetBufferPosition.y });

```

**Review by AJenbo — CHANGES_REQUESTED** [review]
Using Point in scrollrt.cpp allows for the removal of ShiftGrid which i think makes the code a lot more readable, and should generate more optimized code as well. I'll leave that part optional but still made comments in case you are willing to do that also or just for feedback.

Other then that there are just 5 comments that I think should be addressed before this is ready to merge.

File: Source/loadsave.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (AJenbo) [reply Source/loadsave.cpp:760]:
```suggestion
	ReturnLvlPosition.x = file->NextBE<int32_t>();
	ReturnLvlPosition.y = file->NextBE<int32_t>();
```

File: Source/scrollrt.cpp (hunk 16 above)
Comment (StephenCWills) [reply Source/scrollrt.cpp:1547]: Double-check this one. I think it's `DIR_N`.

**Comment by julealgon** [comment]
> Using Point in scrollrt.cpp allows for the removal of ShiftGrid which i think makes the code a lot more readable

Usages can be replaced, but the function cannot be removed yet as there are other usages of it still relying on raw `int`s in a few places.

File: Source/loadsave.cpp [hunk 20]
```diff
@@ -757,8 +757,9 @@ void LoadQuest(LoadHelper *file, int i)
 		file->Skip(1); // Alignment
 	quest._qlog = file->NextBool32();
 
-	ReturnLvlX = file->NextBE<int32_t>();
-	ReturnLvlY = file->NextBE<int32_t>();
+	ReturnLvlPosition = {};
```
Comment (AJenbo) [inline Source/loadsave.cpp:760]: I don't think there is a need for zeroing this out first?

File: Source/scrollrt.cpp [hunk 21]
```diff
@@ -1192,28 +1159,28 @@ void DrawGame(const Surface &fullOut, int x, int y)
...
 		sx -= TILE_WIDTH;
-		ShiftGrid(&x, &y, -1, 0);
+		position += DIR_W;
 		columns++;
 		break;
 	case SDIR_NW:
 		sx -= TILE_WIDTH / 2;
 		sy -= TILE_HEIGHT / 2;
-		x--;
+		position += DIR_NW;
```
Comment (AJenbo) [inline Source/scrollrt.cpp:1174]: Nice now I finally understand this code, it need to shift if moving N or W :D

File: Source/scrollrt.cpp (hunk 16 above)
Comment (AJenbo) [reply Source/scrollrt.cpp]: Good catch

File: Source/scrollrt.cpp (hunk 21 above)
Comment (julealgon) [reply Source/scrollrt.cpp:1174]: At least someone understands it. I don't.

**Review by AJenbo — APPROVED** [review]

**Pull request merged**